Reduce draw calls

kazcw opened the pull request:
### Pull Request Description

Implements #6544 (eliminates 10/42 of the constantly-displayed draw calls).
Fixes #6717. Improves startup CPU time by 5% (250ms, loading Orders on my dev box).

### Important Notes

- Edges: New implementation uses only Rectangle under most conditions.
- Node and action area: Replace some shapes with Rectangle.
- List view: Replace some shapes with Rectangle.
- Display object hierarchy: The lowest-level shape instance types no longer have their own display objects.
- Includes initial support for using `Rectangle` to display triangles.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

kazcw committed:
label

kazcw committed:
listview

kazcw committed:
node

kazcw committed:
action bar

kazcw committed:
edge

kazcw committed:
support inserting before the first partition

kazcw committed:
edge: hover extensions

kazcw committed:
wip edge

kazcw committed:
edge mouse events

kazcw committed:
wip edge

kazcw committed:
SplitArc

kazcw committed:
Ignore points too far from the edge

kazcw committed:
Set node/edge partitions

kazcw committed:
Refactor.

kazcw committed:
Edge target attachments.

kazcw commented:
I'll be testing this some more in the morning (with a Book Club), but it's ready for review.

vitvakatu commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
-            }
+        let output = &frp.private.output;
+        frp::extend! { network
+            // Setters. Changes to these values don't automatically trigger a redraw.
```
I think this should be mentioned in the documentation of the FPR definition.

vitvakatu commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
+            eval frp.set_disabled    ((t) model.set_disabled(*t));
+
+            // Mouse events.
+            eval mouse_move            ((e) model.set_mouse_position_and_redraw(e.client_centered()));
```
We no longer use spacing in FRP code, and it isn't consistent here anyway, so I suggest removing redundant spaces.

vitvakatu commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
+            display_object,
+            source_attached,
+            scene,
+            source_width: default(),
+            source_height: default(),
+            target_position: default(),
+            target_attached: default(),
+            color: default(),
+            sections: default(),
+            split_arc: default(),
+            hover_sections: default(),
+            corner_points: default(),
+            target_attachment: default(),
+            hover_position: default(),
+            disabled: default(),
+            previous_target: default(),
+            previous_is_hoverable: default(),
+            previous_hover_split: default(),
+            previous_color: default(),
+            previous_target_attached: default(),
```
I wonder if using `..default()` is reasonable here.

kazcw replied:
I wish I could do it that way, but I can't `impl Default for EdgeModel` because it must hold a `Scene`, so I can't default-construct an `EdgeModel` object to use for field-update syntax.

wdanilo replied:
We could split this into two structs and deref to the inner struct, making the inner struct Default. I'm not sure though if we want it. Up to you :)

vitvakatu commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,951 @@ use crate::prelude::*;
-                $(output.push(&self.$field);)*
-                output
-            }
+    fn source_x_range(&self) -> f32 {
```
What is `source_x_range`? I'd expect `Range<f32>` as the return type for anything with `range` in the name.

vitvakatu reviewed (commented):
I can also see some changes to the list view. Please mention them in the PR description so QA reviewers know what to check carefully.

kazcw committed:
Review.

kazcw committed:
Rectangle backwards-edge arrows.

kazcw commented:
There's a flicker when the mouse enters/leaves the arc area of a corner, caused by an existing bug. Issue: #6717

kazcw committed:
Fix edge case

kazcw committed:
Fix target attachment coloring

kazcw committed:
Tweak distance computation

kazcw committed:
List view appearance

kazcw committed:
List view event delivery

kazcw committed:
Clean up list view mouse events

kazcw committed:
Fix profiling for microtasks

wdanilo commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,1044 @@ use crate::prelude::*;
+            let target_end = self.with_source_end(snapped);
+            Some(SplitCorner { source_end, target_end, split_arc: None })
+        } else if x_near_vertical && y_along_vertical {
+            // The point is along the vertical line. Snap its x-value, and draw a corner to it.
+            let snapped = Vector2(self.vertical.x(), split_point.y());
+            let source_end = self.with_target_end(snapped);
+            let target_end = self.with_source_end(snapped);
+            Some(SplitCorner { source_end, target_end, split_arc: None })
+        } else if on_arc {
+            // Find the input point's angle along the arc.
+            let offset_from_origin = split_point - origin;
+            let split_angle = offset_from_origin.y().atan2(offset_from_origin.x());
+            // Split the arc on the angle.
```
In the new implementation, a lot of docs from the old implementation are missing, something like this:

```
// node (for example, not to draw the port line above source node).
         //
         // ╭──────────────╮
         // │      ┼ (0,0) │
         // ╰──────────────╯────╮
         //                     │
         //                     ▢ target
```

These docs were very helpful to understand which part of a connection the code relates to. Can we bring them back, please?

kazcw replied:
Added docs with diagrams to the `layout` module, explaining the key concepts of the design.

wdanilo commented on app/gui/view/graph-editor/src/component/edge.rs:
```diff
@@ -4,2033 +4,1044 @@ use crate::prelude::*;
-                fn set_focus_split_angle(&self, angle: f32) {
-                    self.focus_split_angle.set(angle);
-                }
+    fn new_section(&self) -> Rectangle {
+        let new = Rectangle::new();
+        new.set_corner_radius_max();
+        new.set_inset_border(LINE_WIDTH);
```
These functions should have at least small docs. Based on this name, it's hard to understand what a "section" is.

wdanilo reviewed (changes requested):
This is a GIGANTIC change, amazing job! I'm marking this as "request changes" only because:

1. There are missing docs in several places for private functions, whose names are not explanatory enough. Edge drawing is super complex and good docs there will be a big time saver for the future refactoring / improvements.
2. There is no video attached to this PR. If a PR touches visual things, a video showing as many edge cases as possible should be attached.


Also, Kaz, do you think you'd be able to create a debug scene with an edge only? (no nodes). Just an edge between two boxes, where you can drag one of the box? This will be insanely helpful for edge debugging / improving. I always wanted to do it, but it required getting into the edges drawing code - as you know this code well now, this should be doable pretty fast, what do you think about it? :)

wdanilo commented:
Also, Kaz, could you share some stats, please? What is the draw call count drop after this PR? :)

kazcw committed:
Fix render latency

kazcw committed:
Support Sprite without ShapeSystem with SpriteObject

kazcw committed:
fmt

kazcw committed:
Merge remote-tracking branch 'origin/develop' into fast-edges

kazcw committed:
WASM size limit

kazcw committed:
Clean up invalidation.

- `redraw` is performed automatically, at most once.
- 2-phase layout hack is moved into Edge and made a little more
  granular, at least.

kazcw committed:
Edges demo scene.

kazcw committed:
rm extra file

kazcw committed:
Merge remote-tracking branch 'origin/develop' into fast-edges

kazcw commented:
> Also, Kaz, do you think you'd be able to create a debug scene with an edge only? (no nodes). Just an edge between two boxes, where you can drag one of the box? This will be insanely helpful for edge debugging / improving. I always wanted to do it, but it required getting into the edges drawing code - as you know this code well now, this should be doable pretty fast, what do you think about it? :)

Created a `graph_view_edges` debug scene, that shows an edge between draggable dummy nodes. I also cleaned up the FRP *a lot*, which simplified the debug scene and graph-editor.

kazcw commented:
> Also, Kaz, could you share some stats, please? What is the draw call count drop after this PR? :)

Testing Orders, showing just the graph right after opening:
Before (current `develop`):
![image](https://github.com/enso-org/enso/assets/1047859/7aada3c0-cae5-44ae-bdf3-48a5d32c5411)
After (this branch, currently):
![image](https://github.com/enso-org/enso/assets/1047859/695ea4c1-2239-4fae-97c7-b0d3f69e54fc)

Eliminated 10/42 (~24%) of the constantly-displayed draw calls.

kazcw committed:
Refactor.

kazcw committed:
Fix source snap.

kazcw committed:
Avoid visual glitch if internal error occurs.

kazcw committed:
Refactor.

kazcw committed:
Refactor.

kazcw committed:
Clean up accidentally-committed files.

kazcw committed:
Docs.

kazcw committed:
Docs.

kazcw commented:
> 1. There are missing docs in several places for private functions, whose names are not explanatory enough. Edge drawing is super complex and good docs there will be a big time saver for the future refactoring / improvements.
> 2. There is no video attached to this PR. If a PR touches visual things, a video showing as many edge cases as possible should be attached.

Added a video, and a lot more docs. I think I've addressed everything. I also refactored the code a lot, and I think it's easier to read now.

kazcw committed:
Merge remote-tracking branch 'origin/develop' into fast-edges

wdanilo reviewed (changes requested):
Based on the video, I've spotted several regressions:

1. The lower part of the edge is not above the nodes anymore:
    <img width="88" alt="image" src="https://github.com/enso-org/enso/assets/1623053/8911bf37-07b8-4626-8563-be0f8c6acbe0">
2. Edges going up do not have up arrow visible.
3. Edges have different corners than currently. The current behavior of edges was carefully designed to not inference with nodes, the new one is way more curvy and less uniform. Consider the old vs the new one:
     <img width="376" alt="image" src="https://github.com/enso-org/enso/assets/1623053/f6493113-5c8a-48d0-8209-c8aa042b3335"> 
    <img width="414" alt="image" src="https://github.com/enso-org/enso/assets/1623053/b8d74f59-a806-498c-9385-8429c5b24996">
4. I think the behavior of edges close to node edge is different than it was. It was smoothly moving around and the corner radius was depending on the mouse distance:
https://github.com/enso-org/enso/assets/1623053/424ec53a-a2e1-4362-8f32-640c0cc0a781


Basically, we should preserve the look of edges as we have right now because we spent really lot of time designing their current behavior. If we want to change any part of it, let's have a call about it and let's talk about why such a change will introduce better UX to the user. But until we change the design, we need to preserve 100% of the look and feel when doing performance improvements.

wdanilo reviewed (changes requested):
Based on the video, I've spotted several regressions:

1. The lower part of the edge is not above the nodes anymore:
    <img width="88" alt="image" src="https://github.com/enso-org/enso/assets/1623053/8911bf37-07b8-4626-8563-be0f8c6acbe0">
2. Edges going up do not have up arrow visible.
3. Edges have different corners than currently. The current behavior of edges was carefully designed to not inference with nodes, the new one is way more curvy and less uniform. Consider the old vs the new one:
     <img width="376" alt="image" src="https://github.com/enso-org/enso/assets/1623053/f6493113-5c8a-48d0-8209-c8aa042b3335"> 
    <img width="414" alt="image" src="https://github.com/enso-org/enso/assets/1623053/b8d74f59-a806-498c-9385-8429c5b24996">
4. I think the behavior of edges close to node edge is different than it was. It was smoothly moving around and the corner radius was depending on the mouse distance:
https://github.com/enso-org/enso/assets/1623053/424ec53a-a2e1-4362-8f32-640c0cc0a781


Basically, we should preserve the look of edges as we have right now because we spent really lot of time designing their current behavior. If we want to change any part of it, let's have a call about it and let's talk about why such a change will introduce better UX to the user. But until we change the design, we need to preserve 100% of the look and feel when doing performance improvements.

kazcw committed:
Draw detached edge above nodes, except its source.

kazcw committed:
Revert "Avoid visual glitch if internal error occurs."

This reverts commit c512b06f27faec6513854fff5498967b9a204a05.

kazcw committed:
MAX_RADIUS adjustments.

kazcw committed:
Adjust backward-edge arrow threshold.

kazcw committed:
Edge doesn't always enter target node at the top.

kazcw committed:
Docs.

kazcw committed:
Adjust U-bend height.

kazcw committed:
Adjust J0>J1 threshold.

kazcw committed:
Edge exits source before turning up.

kazcw committed:
Adjust max radius.

kazcw committed:
Downward edge originates further inside source.

kazcw committed:
Refactor.

kazcw committed:
Add old edge to demo scene.

kazcw committed:
Upward corners exit before bending.

kazcw committed:
Exact U-bend height.

kazcw committed:
Apply min-approach-height in all layouts.

kazcw committed:
Refactor.

kazcw committed:
Docs.

kazcw committed:
Exact J0>J1 transition point.

kazcw committed:
Exact approach height test.

kazcw committed:
Refactor.

kazcw committed:
Source end of edge starts out horizontal.

kazcw committed:
Single-corner layout radius adjustment.

kazcw committed:
Refactor.

kazcw committed:
Adjust radii in 1st 3-corner Layout.

kazcw committed:
3-corner small-radius height adjustment.

kazcw committed:
Refactor.

kazcw committed:
Fix arc hover edge case.

kazcw committed:
Fix hover edge case.

kazcw committed:
Fix 2px wiggle.

kazcw committed:
Now triangles can be Rectangles too

kazcw committed:
Rectangle arrows.

kazcw committed:
Cleanup.

kazcw committed:
Deviate from old behavior in "shark fin" case.

kazcw committed:
Source end exits node normal to the curve.

kazcw committed:
Adjust a threshold.

kazcw committed:
Fix threshold.

kazcw committed:
Use 3-corner when target on top of node.

kazcw committed:
Make very short edges less sharp.

kazcw commented:
Updated demo; the old edges are in red:

[vokoscreenNG-2023-06-01_16-37-03.webm](https://github.com/enso-org/enso/assets/1047859/c958b522-13ab-457a-9d16-186a32a9ffc0)

Known differences:
- At the beginning of the video, you can see the radius is different in some positions when the target is very close to the source. The old edge snaps abruptly between a sharp corner and a smoother angle. The new edge interpolates between these extremes, avoiding a sudden transition in this case.
- Near the end of the video, when the target is slightly higher than the source, and barely to the side of the source, the old edges form a shark fin shape, with two different radii side by side. The new edges maintain symmetry in this case.

kazcw committed:
Merge remote-tracking branch 'origin/develop' into fast-edges

kazcw committed:
Fix merge conflict.

wdanilo commented on app/gui/view/graph-editor/src/component/edge/layout.rs:
```diff
@@ -0,0 +1,671 @@
+//! (or just one corner end, at an end of an edge). A series of junction points, always alternating
+//! horizontal/vertical, has a one-to-one relationship with a sequence of corners.
+//!
+//! # Partial edges
+//!
+//! Corners are sufficient to draw any complete edge; however, in order to split an edge into a
+//! focused portion and an unfocused portion at an arbitrary location based on the mouse position,
+//! we need to subdivide one of the corners of the edge.
+//!
+//!                  |\
+//!                  | 3
+//!                 /
```
these docs are beautiful ❤️

wdanilo reviewed (approved):
Looks fantastic! Let's remove the old edge impl and let's merge it! Amazing job with fixing the corner cases where the old edges misbehaved!

kazcw committed:
Lint.

kazcw committed:
Merge branch 'develop' into wip/kw/fix-render-latency

kazcw committed:
Improve terminology.

kazcw committed:
Merge remote-tracking branch 'origin/wip/kw/fix-render-latency' into fast-edges

kazcw committed:
Fix missed usage in rename.

kazcw committed:
Tweak layering.

kazcw committed:
Remove old edges from example scene.

kazcw committed:
Merge branch 'develop' into wip/kw/fast-edges

kazcw committed:
Escape diagrams so they don't scare the lexer.

vitvakatu commented:
QA: passed 📗

kazcw committed:
Bump wasm limit.

kazcw committed:
Bump wasm limit.
